internal/record/record.go

```diff
@@ -0,0 +1,123 @@
+package record
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"gopkg.in/yaml.v3"
+
+	"github.com/DanWlker/remind/internal/config"
+	"github.com/DanWlker/remind/internal/data"
+	i_error "github.com/DanWlker/remind/internal/error"
+)
+
+func GetFile() (string, error) {
+	dataFolder, err := data.GetFolder()
+	if err != nil {
+		return "", fmt.Errorf("helper.GetDataFolder: %w", err)
+	}
+
+	fullpath := filepath.Join(dataFolder, config.DefaultDataRecordFullFileName)
+
+	if _, err := os.Stat(fullpath); errors.Is(err, os.ErrNotExist) {
+		if _, err := os.Create(fullpath); err != nil {
+			return "", fmt.Errorf("os.Create: %w", err)
+		}
+		globalRecordEntity, err := CreateNewRecord("")
+		if err != nil {
+			return "", fmt.Errorf("CreateNewRecord: %w", err)
+		}
+
+		if err := SetFileContents([]RecordEntity{globalRecordEntity}); err != nil {
+			return "", fmt.Errorf("SetRecordFileContents: %w", err)
+		}
+	} else if err != nil {
+		return "", fmt.Errorf("os.Stat: %w", err)
+	}
+
+	return fullpath, nil
+}
+
+func GetFileContents() ([]RecordEntity, error) {
+	recordFile, err := GetFile()
+	if err != nil {
+		return nil, fmt.Errorf("GetRecordFile: %w", err)
+	}
+
+	f, err := os.Open(recordFile)
+	if err != nil {
+		return nil, fmt.Errorf("os.Open: %w", err)
+	}
+	defer f.Close()
+
+	var (
+		items []RecordEntity
+		dec   = yaml.NewDecoder(f)
+	)
+
+	if err := dec.Decode(&items); err != nil {
+		return nil, fmt.Errorf("in yaml decoding: %w", err)
+	}
+
+	return items, nil
+}
+
+func SetFileContents(items []RecordEntity) error {
+	recordFile, err := GetFile()
+	if err != nil {
+		return fmt.Errorf("GetRecordFile: %w", err)
+	}
+
+	f, err := os.OpenFile(recordFile, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0o644)
+	if err != nil {
+		return fmt.Errorf("os.OpenFile: %w", err)
+	}
+	defer f.Close()
+
+	enc := yaml.NewEncoder(f)
+	if err := enc.Encode(items); err != nil {
+		return fmt.Errorf("in yaml encoding: %w", err)
+	}
+
+	if err := f.Close(); err != nil {
```

Comment:
Yes it is. If you want a deferred function to affect the outer function's return value, then as you discovered, you need to use named result parameters. (See https://go.dev/wiki/CodeReviewComments#named-result-parameters.) And `errors.Join` is a good way to preserve both the error that made your function return (if any), and the error encountered while cleaning up.

Most of the time I prefer how simple `defer f.Close()` looks, and when the surrounding function returns due to an error, I'm OK losing any error from the `Close` call. But when the function is about to return normally, I prefer to capture any error from `Close`, which is why there's also an explicit (non-deferred) call near the end.

(In that case, the redundant deferred `Close` still runs, but it's a no-op since the file's already closed, and conveniently the "already-closed" error is disregarded.)